Refactor markdown files (#14287)
minetest/minetest: doc/lua_api.md

> # Minetest Lua Modding API Reference

Author: Done!

          `enable_damage`, `creative_mode`, `enable_server`.
    * `map_persistent`: Specifies whether newly created worlds should use
      a persistent map backend. Defaults to `true` (= "sqlite3")

>     * `author`: The author's ContentDB

Author: Re-added the word.

Mods should then be placed in `<worldname>/game/mods/`.

## Modpacks

Mods can be put in a subdirectory, if the parent directory, which otherwise
should be a mod, contains a file named `modpack.conf`.
The file is a key-value store of modpack details.

* `name`: The modpack name. Allows Minetest to determine the modpack name even
  if the folder is wrongly named.
* `title`: A human-readable title to address the modpack. See [Translating content meta](#translating-content-meta).
* `description`: Description of mod to be shown in the Mods tab of the main menu.
* `author`: The author of the modpack. It only appears when downloaded from
  ContentDB.
* `release`: Ignore this: Should only ever be set by ContentDB, as it is an
  internal ID used to track versions.
* `title`: A human-readable title to address the modpack.

> [!NOTE]

> > To support 0.4.x, please also create an empty modpack.txt file.

Author: Fixed section.

It is parsed by the main menu settings dialogue to list mod-specific
settings in the "Mods" category.

### `init.lua`

The main Lua script. Running this script should register everything it
wants to register. Subsequent execution depends on Minetest calling the
registered callbacks.

`minetest.settings` can be used to read custom or existing settings at load

> time, if necessary. (See [`Settings`])

Author: Oops, moved back

to individual clients, or render things that would be too
expensive to compose with `[combine:`.

> [!WARNING]
> Avoid sending large images this way. This is not a replacement

> for asset files, do not use it to do anything that you could instead achieve

Author: Added suggestion.

This makes the item capable of digging nodes that fulfill both of these:

* Have the `crumbly` group

> * Have a `level` group less or equal to `2`\n  [a-z]

Author: Removed.

        2   180    60    20     -     -
        3   180    60    20     -     -

> [!NOTE]
>   * At `crumbly==0`, the node is not diggable.
>   * At `crumbly==3`, the level difference digging time divider kicks in and makes
>    easy nodes to be quickly breakable.
>   * At `level > 2`, the node is not diggable, because it's `level > maxlevel`

# Entity Damage Mechanism

Damage calculation:

```lua
damage = 0
for group in cap.damage_groups do
    armor = object.armor_groups[group] or 0

>     damage = damage
>       + cap.damage_groups[group]

Reviewer: this is pseudo-code so we can keep the concise syntax
```suggestion
    damage += cap.damage_groups[group]
```

Author: Added suggestion.

* Must be used after the `size`, `position`, and `anchor` elements (if present).
* Defines how much space is padded around the formspec if the formspec tries to
  increase past the size of the screen and coordinates have to be shrunk.
* For X and Y, `0.0` represents no padding (the formspec can touch the edge of the
  screen), and `0.5` represents half the screen (which forces the coordinate size
  to `0`). If negative, the formspec can extend off the edge of the screen.
* Defaults to `[0.05, 0.05]`.

### `no_prepend[]`

* Must be used after the `size`, `position`, `anchor`, and `padding` elements
  (if present).
* Disables `player:set_formspec_prepend` from applying to this formspec.

### `real_coordinates[<bool>]`

> [!NOTE]

> > Enable it automatically using formspec version 2 or higher.

Reviewer: ```suggestion
> Enabled automatically when using formspec version 2 or higher.
```

Author: Added suggestion.

* The label formspec element displays the text set in `label`
  at the specified position.
> [!NOTE]
> If the new coordinate system is enabled, labels are positioned
  from the center of the text, not the top.
* The text is displayed directly without automatic line breaking,
  so label should not be used for big text chunks. Newlines can be
  used to make labels multiline.
> [!NOTE]
> With the new coordinate system, newlines are spaced with half a

>   coordinate. With the old system, newlines are spaced 2/5 of an inventory slot.

Reviewer: also missing `>` here and above

Author: I searched for `\n> .*\n[^>*] ` and resolved all of these issues that I could find.

### `hypertext[<X>,<Y>;<W>,<H>;<name>;<text>]`

* Displays a static formatted text with hyperlinks.
> [!WARNING]
> This element is currently unstable and subject to change.
* `x`, `y`, `w` and `h` work as per field
* `name` is the name of the field as returned in fields to `on_receive_fields` in case of action in text.
* `text` is the formatted text using `Markup Language` described below.

### `vertlabel[<X>,<Y>;<label>]`

* Textual label drawn vertically
* `label` is the text on the label
> [!NOTE]
> If the new coordinate system is enabled, vertlabels are positioned

>   from the center of the text, not the left.

Reviewer: and here

* `vector.new(x, y, z)`
* `{x=num, y=num, z=num}` (Even here you are still supposed to use `vector.new`.)

## Compatibility Notes

Vectors used to be defined as tables of the form `{x = num, y = num, z = num}`.
Since Minetest 5.5.0, vectors additionally have a metatable to enable easier use.
> [!NOTE]
> Those old-style vectors can still be found in old mod code. Hence, mod and

> > engine APIs still need to be able to cope with them in many places.

Author: Added suggestion.

The `lang_code` to use for a given player can be retrieved from
the table returned by `minetest.get_player_information(name)`.

> [!NOTE]
> This functionality should only be used for sorting, filtering
or similar purposes. You do not need to use this to get translated strings

> to show up on the client.

Reviewer: seems to be missing `>` here

Author: True, added them

    lacunarity = 2.0,
    flags = "eased",
},

noise_threshold = 1.6
```

> [!WARNING]
> Use this ore type *very* sparingly since it is ~200x more computationally

>   expensive than any other ore.

Reviewer: and here

    * `actor`: `"player:<name>"`, also `"liquid"`.

## Defaults for the `on_place` and `on_drop` Item Definition Functions

* `minetest.item_place_node(itemstack, placer, pointed_thing[, param2, prevent_after_place])`
    * Place item as a node
    * `param2` overrides `facedir` and wallmounted `param2`
    * `prevent_after_place`: if set to `true`, `after_place_node` is not called
      for the newly placed node to prevent a callback and placement loop
    * returns `itemstack, position`
        * `position`: the location the node was placed to. `nil` if nothing was placed.
* `minetest.item_place_object(itemstack, placer, pointed_thing)`
    * Place item as-is
    * returns the leftover itemstack
> [!WARNING]
> This function is deprecated and will never be called.
* `minetest.item_place(itemstack, placer, pointed_thing[, param2])`
    * Wrapper that calls `minetest.item_place_node` if appropriate
    * Calls `on_rightclick` of `pointed_thing.under` if defined instead
    * `param2` overrides facedir and wallmounted `param2`
    * returns `itemstack, position`
        * `position`: the location the node was placed to. `nil` if nothing was placed.
> [!NOTE]

> > `on_rightclick` is not called when wielded item overrides `on_place`

Reviewer: ```suggestion
> `on_rightclick` is not called when the wielded item overrides `on_place`
```

Author: Added suggestion.

    * `step` determines how fast a sound will fade.
      The gain will change by this much per second,
      until it reaches the target gain.
      Note: Older versions used a signed step. This is deprecated, but old
      code will still work. (the client uses abs(step) to correct it)
    * `gain` the target gain for the fade.
      Fading to zero will delete the sound.
> [!NOTE]
> Older versions used a signed step. This is deprecated, but old code will
> still work. (the client uses `abs(step)` to correct it)

Timing
------
## Timing

* `minetest.after(time, func, ...)`: returns job table to use as below.
    * Call the function `func` after `time` seconds, may be fractional
    * Optional: Variable number of arguments that are passed to `func`
    * Jobs set for earlier times are executed earlier. If multiple jobs expire

>       at exactly the same time, then they are executed in registration order.

Reviewer: accidental content removal

Author: Re-added content

    * Clients will attempt to fetch files added this way via remote media,
      this can make transfer of bigger files painless (if set up). Nevertheless
      it is advised not to use dynamic media for big media files.
> [!NOTE]
> If `ephemeral` = false and `to_player` is unset, then the file is added to the media
> sent to clients on startup. This means the media will appear even on old clients if

> > they rejoin the server.

Reviewer: forgot to delete the copy

Author: Fixed notes here

* `minetest.raillike_group(name)`: returns a rating
    * Returns rating of the connect_to_raillike group corresponding to name
    * If name is not yet the name of a connect_to_raillike group, a new group
      ID is created, with that name.
* `minetest.get_content_id(name)`: returns an integer
    * Gets the internal content ID of `name`
* `minetest.get_name_from_content_id(content_id)`: returns a string
    * Gets the name of the content with that content ID

> * `minetest.parse_json(string[, nullvalue])`: returns parsed datatype

Reviewer: ```suggestion
* `minetest.parse_json(string[, nullvalue])`: returns parsed data
```

Author: Added suggestion.

    * styled: Outputs in a human-readable format if this is set, defaults to
      false.
    * Unserializable things like functions and userdata will cause an error.
    * Example: `write_json({10, {a = false}})`,
      returns `'[10, {"a": false}]'`
> [!WARNING]

> > JSON is more strict than the Lua table format.

Reviewer: ```suggestion
> JSON is stricter than the Lua table format:
```

Author: Added suggestion.

* `set_string(key, value)`: Value of `""` will delete the key.
* `get_string(key)`: Returns `""` if key not present.
* `set_int(key, value)`

>     * The range for the value is system-dependent (usually four bytes).

Reviewer: this isn't any clearer, just different.

      The value will be converted into a string when stored.
* `get_int(key)`: Returns `0` if key not present.
* `set_float(key, value)`

>     * The range for the value is system-dependent (usually double precision).

Reviewer: 32 bits is "single precision"

Author: Removed both bit rewordings.

To change position or rotation call `set_attach` again with the new values.

> [!NOTE]
> Just like model dimensions, the relative position in `set_attach` must

>   be multiplied by 10 compared to world positions.

Reviewer: and here

    * Sets the position of the object.
    * No-op if object is attached.
    * `pos` is a vector `{x=num, y=num, z=num}`
* `add_pos(pos)`:
    * Changes position by adding to the current position.

>     * No-op if object is attached.

Reviewer: rebase mistake

Author: Re-added section.

    * `vel` is a vector, e.g. `{x=0.0, y=2.3, z=1.0}`
    * In comparison to using get_velocity, adding the velocity and then using
      set_velocity, add_velocity is supposed to avoid synchronization problems.

>       Additionally, players also do not support set_velocity.

Reviewer: I thought function names were supposed to be in code formatting?

Author: You are right, changed

    object.
* `set_detach()`: Detaches object. No-op if object was not attached.
* `set_bone_position([bone, position, rotation])`
    * Shorthand for `set_bone_override(bone, {position = position, rotation = rotation:apply(math.rad)})` using absolute values.
> [!WARNING]
> **Deprecated**, use `set_bone_override` instead.

> [!NOTE]
> Rotation is in degrees, not radians.
* `get_bone_position(bone)`: returns the previously set position and rotation of the bone
    * Shorthand for `get_bone_override(bone).position.vec, get_bone_override(bone).rotation.vec:apply(math.deg)`.
> [!WARNING]
> **Deprecated**, use `get_bone_override` instead.

> [!NOTE]
> Returned rotation is in degrees, not radians.
* `set_bone_override(bone, override)`
    * `bone`: string

>     * `override`: `position = property, rotation = property, scale = property}` or `nil`

Reviewer: missing `{`
also it's more readable with spacing

Author: Added back the `{`, but I will not be putting the spacing back. This is one of the points of this PR

    * `framelength`: Time per animated frame in seconds, default: `0.2`
    * `select_x_by_camera`: Only for visual = `sprite`. Changes the frame `x`
      position according to the view direction. default: `false`.
        * First column: subject facing the camera
        * Second column: subject looking to the left
        * Third column: subject backing the camera
        * Fourth column: subject looking to the right
        * Fifth column:  subject viewed from above
        * Sixth column:  subject viewed from below
* `get_luaentity()`:
	* Returns the object's associated luaentity table, if there is one
	* Otherwise returns `nil` (e.g. for players)
* `get_entity_name()`:
  * **Deprecated**: Will be removed in a future version,
    use `:get_luaentity().name` instead.

#### Player Only (No-Op for Other Objects)

> * `get_player_name()`: returns `""` if is not a player

Reviewer: content accidentally removed.

Author: Re-added.

        * `fog`: A table with following optional fields:
            * `fog_distance`: integer, set an upper bound the client's viewing_range (inluding range_all).
              By default, fog_distance is controlled by the client's viewing_range, and this field is not set.
              Any value >= 0 sets the desired upper bound for the client's viewing_range and disables range_all.
              Any value < 0, resets the behavior to being client-controlled.
              (default: -1)
            * `fog_start`: float, override the client's fog_start.
              Fraction of the visible distance at which fog starts to be rendered.
              By default, fog_start is controlled by the client's `fog_start` setting, and this field is not set.
              Any value between [0.0, 0.99] set the fog_start as a fraction of the viewing_range.
              Any value < 0, resets the behavior to being client-controlled.
              (default: -1)
            * `fog_color`: ColorSpec, override the color of the fog.
              Unlike `base_color` above this will apply regardless of the skybox type.
              (default: `"#00000000"`, which means no override)
> [!WARNING]

> > The darkening of the ColorSpec is subject to change.

Reviewer: "the ColorSpec"
which?

Author: Specified `sky_color`.

* `set_lighting(light_definition)`: sets lighting for the player
    * Passing no arguments resets lighting to its default values.
    * `light_definition` is a table with the following optional fields:
        * `saturation` sets the saturation (vividness; default: `1.0`).
          values > 1 increase the saturation

>           values in [0,1) decrease the saturation

Reviewer: if the default is 1.0 then `[0,1)` must be incorrect

Author: Fixed this section.

    * `mean = (max - min) / 2`, and
    * `variance = (((max - min + 1) ^ 2) - 1) / (12 * num_trials)`
    * Increasing `num_trials` improves accuracy of the approximation
* `get_state()`: return generator state encoded in string

> * `set_state(state_string)`: restore generator state from encoded string

Reviewer: rebase mistake

Author: Re-added.

    * align style determines whether the texture will be rotated with the node
      or kept aligned with its surroundings. "user" means that client
      setting will be used, similar to `glasslike_framed_optional`.
    * scale is used to make texture span several (exactly `scale`) nodes,
      instead of just one, in each direction. Works for world-aligned
      textures only.
      Note that as the effect is applied on per-mapblock basis, `16` should
      be equally divisible by `scale` or you may get wrong results.
> [!NOTE]

> > Supported by solid nodes and nodeboxes only.

Reviewer: context lost

Author: Fixed context.
